### ISSUE TITLE
Modelica.Blocks.Discrete.FirstOrderHold has bugs

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#774]
**Reported by jriel on 6 Jul 2012 17:36 UTC**
Simulate the following and note the non-continuous output at FoH.y:
```
model Main
    import Modelica.Blocks.Sources;
    import Modelica.Blocks.Discrete;
    Discrete.FirstOrderHold  FoH;
    Sources.Sine             Sine;
equation
    connect(Sine.y, FoH.u);
    annotation(probes={FoH.u, FoH.y});
end Main;
```

There are a few bugs in the design of the FirstOrderHold, the most serious being that there is no sampler at the input.  Possibly that is by design, however, the doc-string suggests it should work with a continuous-time input signal.  Here is my redesign of the device,
```
  block FirstOrderHold "First order hold of a sampled-data system"
      extends Interfaces.DiscreteSISO;
  protected
    discrete Real u1, u2, c;
    Modelica.SIunits.Time tSample;
  equation
      when sampleTrigger then
	  u1 = pre(u);
	  u2 = pre(u1);
	  tSample = time;
	  c = if firstTrigger then 0 else (u1-u2)/samplePeriod;
    end when;
    y = u2 + c*(time - tSample);
    annotation (
      Icon(coordinateSystem(
          preserveAspectRatio=true,
          extent={{-100,-100},{100,100}},
          grid={1,1}), graphics={Line(points={{-79,-41},{-59,-33},{-40,1},{-20,
                9},{0,63},{21,20},{41,10},{60,20}}, color={0,0,127}), Line(
              points={{60,19},{81,10}}, color={0,0,255})}),
      Documentation(info="<HTML>
<p>
The output signal is the extrapolation through the
values of the last two sampled input signals.
</p>
</HTML>
"));
  end FirstOrderHold;
```

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/774

**Comment by modelica-trac-importer** [comment]
**Comment by jriel on 6 Jul 2012 18:32 UTC**
On reflection, the line `u1 = pre(u);` in my proposed fix should be `u1 = u;`.

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 20 Dec 2012 13:45 UTC**

**Comment by modelica-trac-importer** [comment]
**Changelog removed by dietmarw on 20 Dec 2012 13:45 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 2 Jan 2013 18:11 UTC**
Fixed in 3b12c8237ee99dec26c5370dfa7c9498aa2b88ea:

Fixed the wrong code along the suggested implementation and added missing initial conditions.
